opalog/parser_test.go

```diff
@@ -6,46 +6,99 @@ package opalog
 
 import (
 	"fmt"
+	"reflect"
 	"testing"
 )
 
 var _ = fmt.Printf
 
-func testTermEqual(t *testing.T, x *Term, y *Term) {
+func TestScalarTerms(t *testing.T) {
+	assertParseOneTerm(t, "null", "null", reflectTerm(nil))
+	assertParseOneTerm(t, "true", "true", reflectTerm(true))
+	assertParseOneTerm(t, "false", "false", reflectTerm(false))
+	assertParseOneTerm(t, "integer", "53", reflectTerm(53))
+	assertParseOneTerm(t, "integer2", "-53", reflectTerm(-53))
+	assertParseOneTerm(t, "float", "16.7", reflectTerm(16.7))
+	assertParseOneTerm(t, "float2", "-16.7", reflectTerm(-16.7))
+	assertParseOneTerm(t, "exponent", "6e7", reflectTerm(6e7))
+	assertParseOneTerm(t, "string", "\"a string\"", reflectTerm("a string"))
+	assertParseOneTerm(t, "string", "\"a string u6abc7def8abc0def with unicode\"",
+		reflectTerm("a string u6abc7def8abc0def with unicode"))
+
+	assertParseOneTermFail(t, "hex", "6abc")
+	assertParseOneTermFail(t, "non-string", "'a string'")
+	assertParseOneTermFail(t, "non-number", "6zxy")
+	assertParseOneTermFail(t, "non-number2", "6d7")
+	assertParseOneTermFail(t, "non-number3", "6\"foo\"")
+	assertParseOneTermFail(t, "non-number4", "6true")
+	assertParseOneTermFail(t, "non-number5", "6false")
+	assertParseOneTermFail(t, "non-number6", "6[null, null]")
+	assertParseOneTermFail(t, "non-number7", "6{\"foo\": \"bar\"}")
+}
+
+func TestVarTerms(t *testing.T) {
+	assertParseOneTerm(t, "var", "foo", reflectTerm(NewVar("foo")))
+	assertParseOneTerm(t, "var", "foo_bar", reflectTerm(NewVar("foo_bar")))
+	assertParseOneTerm(t, "var", "foo0", reflectTerm(NewVar("foo0")))
+
+	assertParseOneTermFail(t, "non-var", "foo-bar")
+	assertParseOneTermFail(t, "non-var2", "foo-7")
+}
+
+func TestObjectWithScalars(t *testing.T) {
+	assertParseOneTerm(t, "number", "{\"abc\": 7, \"def\": 8}", reflectTerm(map[string]int{"abc": 7, "def": 8}))
+	assertParseOneTerm(t, "bool", "{\"abc\": false, \"def\": true}", reflectTerm(map[string]bool{"abc": false, "def": true}))
+	assertParseOneTerm(t, "string", "{\"abc\": \"foo\", \"def\": \"bar\"}", reflectTerm(map[string]string{"abc": "foo", "def": "bar"}))
+	assertParseOneTerm(t, "mixed", "{\"abc\": 7, \"def\": null}", reflectTerm(map[string]interface{}{"abc": 7, "def": nil}))
+}
```

Comment:
What do you think about supporting objects where the key is a number or a boolean?  I know the spec doesn't allow it, but if other json parsers handle it, we probably should too.